marconi-sidechain/src/Marconi/Sidechain/Concurrency.hs

```diff
@@ -0,0 +1,107 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Marconi.Sidechain.Concurrency (
+  HandledAction (Handled, Unhandled),
+  raceSignalHandled_,
+  raceSignalHandled,
+) where
+
+import Control.Concurrent.Async (race)
+import Control.Concurrent.MVar (newEmptyMVar, putMVar, tryTakeMVar)
+import Control.Exception (Exception, catch, throwIO)
+import Control.Monad (void)
+import Data.Void (Void)
+import Marconi.Sidechain.Error (HasExitCode (toExitCode), signalToExit, withSignalHandling)
+import System.Exit (ExitCode (ExitFailure), exitWith)
+
+-- * Graceful exception handling after SIGINT or SIGTERM
+
+{- These purpose of these functions is as follows:
+
+  If we try to naiively exit with a different exit code than that of the received signal, our new
+	exit code will be ignored, We can work around this by using `installHandler`; catching the
+	signal and setting our own exit code.
+
+  We can use exceptions to map our domain failures to these exit codes.
+
+  To do this, we use an MVar to hold any overwritten exit code. The exit code is overwritten
+  by a handler function, which catches exceptions of the desired type. The type is then
+  pattern matched and the exit code is chosen based on the contructor.
+
+  If the program is exiting due to a SIGINT or SIGTERM and if we've got an MVar in our result,
+  we set the exit code to whatever it specifies. If either of the previous points aren't true,
+  we exit with whatever signal (SIGINT or SIGTERM) requested shutdown. -}
+
+{- | A data type representing an action with explicit instructions as to whether concurrent
+		 operations should attempt to map any exceptions it throws to Unix exit codes.
+
+		It has two type parameters:
+
+		* 'e' the exception type, 'Void' in the unhandled case
```

Comment:
I think Haddock don't like single quotes for type parameters, as it tries to resolve them to a known function. You probably want to use `@a@` and `@e@`.